src/app/pages-block/pages-block.component.ts

```diff
@@ -1,45 +1,100 @@
 
   @Output() addButtonClicked: EventEmitter<void> = new EventEmitter()
 
+  showLoadMore: boolean = true;
   courses: ICoursePage[] = [];
+  numberOfCourses: number = 3;
+  private currentSubscribes: Subject<void> = new Subject<void>();
 
-  constructor(private coursesPagesService: CoursesService) { }
+  constructor(private coursesPagesService: CoursesService, private loadingService: LoadingService) { }
 
   ngOnInit(): void {
-    this.courses = this.coursesPagesService.getCoursesList()
+
+    setTimeout(() => this.refreshCourse(), 0);
+    this.searchFunction();
+  }
+
+  searchFunction(): void {
+    const searchBox = document.getElementById('search-box') as HTMLInputElement;
+    if (searchBox) {
+      let searchData$ = fromEvent(searchBox, 'input').pipe(
+        map(elem => (elem.target as HTMLInputElement).value),
+        filter(text => text.length > 2 || text === ''),
+        debounceTime(500),
+        distinctUntilChanged(),
+      );
+
+      searchData$.pipe(
+        takeUntil(this.currentSubscribes)
+      ).subscribe((inputData: string) => {
+        if (inputData) {
+          this.coursesPagesService.getFilteredList(inputData.toLowerCase()).pipe(
+            takeUntil(this.currentSubscribes)
+          ).subscribe((courseData) => this.courses = courseData);
+        } else {
+          this.refreshCourse();
+        }
+      });
+    }
   }
 
   deleteComponent(id: string): void {
     if (confirm('Do you really want to delete this course? Yes/No')) {
-      this.courses = this.coursesPagesService.deleteCourse(id)
+      this.coursesPagesService.deleteCourse(id).pipe(
+        takeUntil(this.currentSubscribes)
+      ).subscribe(() => {
+        this.refreshCourse();
+      })
     }
   }
 
   loadNewCourses(): void {
-    console.log('here is come action');
+    this.numberOfCourses += 3;
+    this.refreshCourse();
   }
 
   changeRate(course: ICoursePage): void {
     course.topRated = !course.topRated;
-    this.coursesPagesService.updateCourse(course);
+    this.loadingService.setValue(true);
+    this.coursesPagesService.updateCourse(course).pipe(
+      takeUntil(this.currentSubscribes)
+    ).subscribe();
+    this.loadingService.setValue(false);
   }
 
-  findClick(inputData: string): void {
-    if (inputData !== '') {
-      this.courses = this.courses.filter(course => course.title.toLowerCase().includes(inputData.toLowerCase()))
-    } else {
-      this.courses = this.coursesPagesService.getCoursesList();
-    }
+  refreshCourse(): void {
+    this.loadingService.setValue(true);
+    let currentNumberOfCourses = this.courses.length;
+    let newNumberOfCourses;
+
+    this.coursesPagesService.getCoursesList(this.numberOfCourses).pipe(
+      takeUntil(this.currentSubscribes)
+    ).subscribe((courseData) => {
+      this.courses = courseData;
+      this.loadingService.setValue(false);
+      newNumberOfCourses = this.courses.length;
+      if (currentNumberOfCourses === newNumberOfCourses) {
```

Comment:
compare directly courseData.length and this.courses.length
and move IF statement at the start of subscription before you assign new value to this.courses